internal/controller/compute/statefulserverset/datavolume_controller.go

```diff
@@ -0,0 +1,174 @@
+
+	"github.com/ionos-cloud/crossplane-provider-ionoscloud/apis/compute/v1alpha1"
+	"github.com/ionos-cloud/crossplane-provider-ionoscloud/internal/controller/serverset"
+	"github.com/ionos-cloud/crossplane-provider-ionoscloud/internal/controller/volumeselector"
+	"github.com/ionos-cloud/crossplane-provider-ionoscloud/pkg/kube"
+)
+
+type kubeDataVolumeControlManager interface {
+	Create(ctx context.Context, cr *v1alpha1.StatefulServerSet, replicaIndex, volumeIndex int) (v1alpha1.Volume, error)
+	Get(ctx context.Context, volumeName, ns string) (*v1alpha1.Volume, error)
+	Delete(ctx context.Context, name, namespace string) error
+	Ensure(ctx context.Context, cr *v1alpha1.StatefulServerSet, replicaIndex, version int) error
+}
+
+// kubeDataVolumeController - kubernetes client wrapper  for server resources
+type kubeDataVolumeController struct {
+	kube client.Client
+	log  logging.Logger
+}
+
+// Create creates a volume CR and waits until in reaches AVAILABLE state
+func (k *kubeDataVolumeController) Create(ctx context.Context, cr *v1alpha1.StatefulServerSet, replicaIndex, volumeIndex int) (v1alpha1.Volume, error) {
+	name := generateNameFrom(cr.Name, volumeselector.ResourceDataVolume, replicaIndex, volumeIndex)
+	k.log.Info("Creating Data Volume", "name", name)
+
+	createVolume := fromSSSetToVolume(cr, name, replicaIndex, volumeIndex)
+	if err := k.kube.Create(ctx, &createVolume); err != nil {
+		return v1alpha1.Volume{}, err
+	}
+	if err := kube.WaitForResource(ctx, kube.ResourceReadyTimeout, k.isAvailable, name, cr.Namespace); err != nil {
+		return v1alpha1.Volume{}, err
+	}
+	// get the volume again before returning to have the id populated
+	kubeVolume, err := k.Get(ctx, name, cr.Namespace)
+	if err != nil {
+		return v1alpha1.Volume{}, err
+	}
+	k.log.Info("Finished creating Data Volume", "name", name)
+
+	return *kubeVolume, nil
+}
+
+// Get - returns a volume kubernetes object
+func (k *kubeDataVolumeController) Get(ctx context.Context, volumeName, ns string) (*v1alpha1.Volume, error) {
+	obj := &v1alpha1.Volume{}
+	err := k.kube.Get(ctx, types.NamespacedName{
+		Namespace: ns,
+		Name:      volumeName,
+	}, obj)
+	return obj, err
+}
+
+// Delete - deletes the datavolume k8s client and waits until it is deleted
+func (k *kubeDataVolumeController) Delete(ctx context.Context, name, namespace string) error {
+	condemnedVolume, err := k.Get(ctx, name, namespace)
+	if err != nil {
+		return err
+	}
+	if err := k.kube.Delete(ctx, condemnedVolume); err != nil {
+		return fmt.Errorf("error deleting data volume %w", err)
+	}
+	return kube.WaitForResource(ctx, kube.ResourceReadyTimeout, k.isDataVolumeDeleted, condemnedVolume.Name, namespace)
+}
+
+// isAvailable - checks if a volume is available
+func (k *kubeDataVolumeController) isAvailable(ctx context.Context, name, namespace string) (bool, error) {
+	obj, err := k.Get(ctx, name, namespace)
+	if err != nil {
+		if apiErrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
```

Comment:
This can be written also like this:
```bash
if err != nil && apiErrors.IsNotFound(err) {
	return false, nil
} else if err != nil {
	return false, err
}
```

Reply:
we decided to go with what was originally written